tests/integration/conftest.py

```diff
@@ -33,30 +33,56 @@ def operator_env(operator_kubeconfig_fp):
 
 
 @pytest.fixture(scope="session", autouse=True)
-def operator(k8s_namespace, operator_env):
+def operator(k8s_namespace, operator_env, k8s_amalthea_api):
+    stdout = open(".kopf.stdout.txt", "w+b")
+    stderr = open(".kopf.stderr.txt", "w+b")
     p = Popen(
         args=f"kopf run -n {k8s_namespace} --verbose kopf_entrypoint.py",
-        stdout=sys.stdout,
-        stderr=sys.stderr,
+        stdout=stdout,
+        stderr=stderr,
         shell=True,
         env={
             **os.environ,
             **operator_env,
         },
         bufsize=0,
+        # NOTE: os.setpgrp makes it so that Ctrl-C is ignored.
+        # This is needed because Ctrl-C gets picked up if the tests are stopped manually
+        # But stopping the operator the instant the tests are stopped results in
+        # incomplete cleanup of fixtures and leftover k8s resources
+        preexec_fn=os.setpgrp,
     )
-    # Give time for the operator to fully start up, call to Popen does not block
-    sleep(2)
+    # INFO: Give time for the operator to fully start up, call to Popen does not block
+    sleep(10)
     yield p
 
-    # We run the KopfRunner again for a short moment to remove all finalizers
-    # and allow cleanup.
-    sleep(10)
-    p.terminate()
-    try:
-        p.wait(timeout=30)
-    except TimeoutExpired:
-        p.kill()
+    # NOTE: Keep the kopf runner (i.e. operator) going until all sessions are cleaned up
+    sessions = k8s_amalthea_api.get(namespace=k8s_namespace)
+    print("Checking for the number of active sessions")
+    while len(sessions.items) > 0:
+        sessions = k8s_amalthea_api.get(namespace=k8s_namespace)
+        print(f"{len(sessions.items)} active sessions found")
+        sleep(10)
+    # NOTE: Ctrl-C is ignored - this is only way to stop operator
+    p.kill()
+    # INFO: Extract and post logs from controller - in most cases controller writes
+    # to standard error regardless of whether erors were present or not
+    stdout.seek(0)
+    stderr.seek(0)
+    stdout_content = stdout.read()
+    stderr_content = stderr.read()
+    Path(".kopf.stdout.txt").unlink(missing_ok=True)
+    Path(".kopf.stderr.txt").unlink(missing_ok=True)
```

Comment:
but to be honest this is why i prefer context-managers, i.e. `with open(...) as....` blocks because no matter what that should cleanup the open file-descriptors; because in case an exception is raised somewhere above those FDs gonna be open till the process actually quits.

Reply:
Yeah that is kind of messy. I wasn't sure if you can have a context manager that spans and should continue to keep the file open even after the `yield` statement. Either way I made the file a fixture using the `tempfile` package. And added a context manager. This way when the fixture is cleaned up and the file is closed the file is automatically deleted.